crates/terminal/src/terminal.rs

```diff
@@ -2096,13 +2100,30 @@ impl Terminal {
         if let Some(tx) = &self.completion_tx {
             tx.try_send(e).ok();
         }
-        if let Some(e) = e {
-            self.child_exited = Some(e);
-        }
+
         let task = match &mut self.task {
             Some(task) => task,
             None => {
-                if self.child_exited.is_none_or(|e| e.code() == Some(0)) {
+                // Always record the latest exit status.
+                if let Some(e) = e {
+                    self.child_exited = Some(e);
+                }
+
+                let should_close = if error_code.is_none() {
+                    let is_interactive_shell =
+                        !matches!(self.template.shell, Shell::WithArguments { .. });
```

Comment:
This doesn't work. Either of the three variants could be an interactive shell. A better approach i think would be to check if `self.task.is_none()`, then we are definitely a shell session and not a task